File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:

        if not height: return 0;

        l, r =0, len(height)-1;
        leftMax ,rightMax = height[l],height[r]
        water = 0

        while(l<r):
            if(leftMax<rightMax):
                l += 1
                leftMax = max(leftMax,height[l])
                water+=leftMax -height[l]
            else: 
                r -= 1
                rightMax = max(rightMax,height[r])
                water += rightMax - height[r]
        
        return water

File: 0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let water = 0;
    let left = 0;
    let right = height.length-1
    let maxLeft= height[left];
    let maxRight = height[right];
    while(left<right){
        if(maxLeft<maxRight){
            left++
            maxLeft = Math.max(maxLeft, height[left])
            water += maxLeft - height[left]
        }else {
            right--;
            maxRight = Math.max(maxRight, height[right])
            water += maxRight - height[right]
        }
    }
    return water;
};